README.txt
==================
About psicons.core
==================

.. |psicons| replace:: *psicons*


Background
----------

Scientific analysis can be problematic:

* It may involve multiple steps, each using the results of the previous stage. 
  Making a mistake often means repeating the whole series for safety.

* Sometimes analysis chains have to be repeated on different datasets.
  Sometimes, even within a single analysis, the same manipulation or test has
  to be repeated with slightly different parameters.

* Even immediately after the fact, it's easy to forget what was done. 9 months
  later when responding to a referee's report, it may be impossible.

* Collaborators, clients or bosses may demand accountability.

* With long but routine tasks, it's easy to get bored and make mistakes.

It this light, |psicons| is a quick and dirty hack to subvert the scons build 
system for scientific analysis. Every stage of analysis is a command-line call 
to a script or executable that takes *inputs* and produces *outputs*. When 
scons is called, dependencies between outputs and inputs are tested and only 
those stages are run that are necessary to update outputs. In addition, the 
exact sequence of analyses is recorded by the build file. 

In summary, |psicons| provides:

* Repeatability: running a build file, reruns the same analysis

* Reproducibility: the build file (and custom scripts) document the steps of 
  the analysis
* Minimization of effort: if inputs or analysis steps are changed, only the 
  necessary (dependent) steps of the analysis are rerun
  
* Mistake-resistant: errors don't derail analysis due to reproducibility ("what 
  did I do") and minimization of effort (only dependent steps are repeated)

* Programmability: analysis tasks may be constructed with programatically
  ("repeat the analysis across this parameter range")


Status
------

*psicons* is very much a hack-and-see project, having been produced in the 
aftermath of the 2009 H1N1 pandemic from the need for complex processing of 
large amounts of sequence data using ad-hoc scripts and formats. It worked 
well in that limited role, but is still an early release exploring the 
approach. Functionality is limited and the API may change. There are other 
more developed (and more specialised) alternatives. Comment is invited.


Installation
------------

The simplest way to install *psicons* is via ``easy_install`` [setuptools]_ or 
an equivalent program::

	% easy_install psicons.core

Alternatively the tarball can be downloaded, unpacked and ``setup.py`` run::

	% tar zxvf psicons-core.tgz
	% cd psicons-core
	% python set.py install

It should work with just about any  version of Python.

*psicons* requires that scons is installed, which is where things get tricky.
Scons by default installs itself in a sandboxed way with multiple versions
living side-by-side and thus being non-importable. Of course, |psicons| needs
to use the scons library, so a conventional installation must be forced. 
Download the scons tarball, unpack it and install it like thus::

	% python setup.py install --standard-lib


Using psicons
-------------

A full API is included in the source distribution.


Examples
~~~~~~~~

Psicons works just like scons. In fact, it is scons. More details are available 
elsewhere but briefly, you run scons like this::

	# look for a build file called "Sconstruct" by default
	% scons 
	# looks for a named build file
	% scons -f mybuildfile

This causes scons to execute the build file, which is just a Python script, 
defining a series of tasks or *commands*::

	# an scons build file
	# some necessary administration - set up the build environment
	env = Environment()
	
	# compile two libraries and then combine into one program
	first_libs = env.Object ('hello.c', CCFLAGS='-DHELLO')
	second_libs = env.Object ('goodbye.c', CCFLAGS='-DGOODBYE')
	env.Program (first_libs + second_libs)

The first time this file is executed, the first two commands build libraries, 
while the third combines the libraries into a single executable. Dependencies 
between the steps are automatically tracked: should one of the original source 
files be changed (e.g. hello.c), when the file is rerun only the steps 
"downstream" of it (e.g. recompilation of the first library, and the final 
linking) are rerun. 

Scons has a large number of commands for all sorts of software builds. Psicons 
adds two new commands, so that local scripts or external programs can be used 
to in a build. In this way, complex multi-step analyses can be constructed from 
a series of interdependent commands, that "build" intermediate data and final 
results::

	from psicons.core import *
	env = Environment()
	
	# call a local script
	IN_DATA = 'jg_08-10_2010.csv'
	CLEAN_DATA = 'jg_08-10_2010-cleaned.csv'
	make_clean_data = Script (env, 'clean_seqs.py',
		args = ['--save-as', CLEAN_DATA],
		infiles = [IN_DATA],
		output = CLEAN_DATA,
	)
	
	# call an external command
	EPI_DATA = 'jg-types.txt'
	RESULT_DATA = 'results.tab'
	type_data = External (env, 'treemaker',
		args = ['--save-as', RESULT_DATA],
		infiles = [CLEAN_DATA, EPI_DATA],
		output = [RESULT_DATA],
	)

The interfaces of these two commands are similar:

* what is being called?
* what inputs does it use (depend on)?
* what outputs does it produce?

When scons is run on this build file, it calls the script 'clean_seqs.py' on 
`IN_DATA` to produce `CLEAN_DATA`. Then the external program `treemaker` is 
called on `CLEAN_DATA` and `EPI_DATA` to produce `RESULT_DATA`. Should 
`EPI_DATA` be edited, when scons is called again, only the second external 
step will be run again as the first step and it's results is still up to date. 
Thus:

* Analyses may be run (and rerun) easily

* If data changes (or scripts change - bug fixes), only the necessary steps are 
  rerun
  
* The actions taken are recorded in the build file

To ease renaming intermediate or output files in a rational way, *psicons* 
offers a few utility functions for interpolating file names from parameters. To 
illustrate::

	# generate a new string from a template
	>>> d = {'foo': '123', 'bar': '456'}
	>>> interpolate ('ab{foo}cd{bar}ef', d)
	'ab123cd456ef'
	# name new file name from old by adding suffix to name 
	>>> interpolate_from_path ('mydata.csv', '{stem}-cleaned{ext}')
	'mydata-cleaned.csv'


Limitations
-----------

Certainly, far, far more complicated reproducibility tools are out there (see `
here <http://csdl2.computer.org/comp/mags/cs/2009/01/mcs2009010005.pdf>`__) but 
many are based around certain disciplines (e.g. geophysics, computational math), 
require working through web interfaces or using very standard sets of analysis 
tools. *psicons* is written from the point of view of a bioinformatician doing 
sequence and phylogenetic analysis, working on the commandline using a lot of 
custom scripts and an endlessly changing lineup of supplied tools. As sometimes 
happens, other tools didn't fit, so I wrote one that did. 

As with many quick hack tools, documentation is currently a bit thin.

The need for a modified scons installation is a blemish. Future versions of
|psicons| may need to directly incorporate scons for ease of installation.

Clearly, a set of standard tools for extracting, transforming and plotting 
data would be a powerful addition to *psicons*. This doesn't exist as yet.

The process of installing a new tool or module for use by SCons is fiddly
[scons-tools]_, involves copying libraries in a tool directory, registering 
their use in build files, voids the ease of using ``easy_install`` and makes 
development a pain. Thus, the additional "commands" are real scons commands, so
much as functions that generate commands. But they are easy to use.


Credit
------

Thanks to the architects of Scons, of course.

While this project was started before encountering Madagascar [madagascar]_, it 
has inevitably shaped development. It's a remarkably powerful system, although 
ill-suited to my current purposes. You should check it out.

Only when writing this document did I become aware of sconstools [sconstools]_,
which seems to be following exactly the same direction as |psicons|.


References
----------

.. [psicons-home] `psicons home page <http://www.agapow.net/software/psicons-core>`__

.. [psicons-pypi] `psicons on PyPi <http://pypi.python.org/pypi/psicons-core>`__

.. [setuptools] `Setuptools & easy_install <http://packages.python.org/distribute/easy_install.html>`__

.. [psicons-github] `psicons on github <https://github.com/agapow/psicons.core>`__

.. [scons] `Scons <http://www.scons.org>`__

.. [madagascar] `Madagscar and Scons for reproducibility <http://reproducibility.org/wiki/Reproducible_computational_experiments_using_SCons>`__

.. [scons-custom] `Where To Put Your Custom Builders and Tools <http://scons.org/doc/production/HTML/scons-user/x3697.html>`__

.. [sconstools] `Sconstools <http://code.google.com/p/sconstools/>`__



docs/HISTORY.txt
=========
Changelog
=========

0.2dev (29072011)
-----------------

- First widely released version
- Added documentation


0.1dev (unreleased)
-------------------

- Initial release


psicons/core/impl.py
#! /usr/bin/env python
# -*- coding: utf-8 -*-
"""
Internal implementation utilities and details.

This module contains various odds and ends to make development easier. None of
the code within should be relied upon as it is subject to change at a whim.
"""

__docformat__ = 'restructuredtext en'


### IMPORTS ###

import types

__all__ = [
	'make_list',
]


### CONSTANTS & DEFINES ###

### IMPLEMENTATION ###	

def make_list (x):
	"""
	If this isn't a list, make it one.
	
	:Parameters:
		x : list, tuple, other
			a sequence, or a single element to be placed in a sequence
			
	:Returns:
		Either the original a parameter if a sequence, or the parameter placed in
		a list.
	
	Syntactic sugar for allowing method calls to be single elements or lists of
	elements.
	
	For example::
	
		>>> make_list (1)
		[1]
		>>> make_list ('1')
		['1']
		>>> make_list ([1, 2])
		[1, 2]
		>>> make_list ((1, 2))
		(1, 2)
		
	"""
	# TODO: should be a more general way of doing this
	if (type (x) not in (types.ListType, types.TupleType)):
		x = [x]
	return x


## DEBUG & TEST ###

if __name__ == "__main__":
	import doctest
	doctest.testmod()



### END #######################################################################


psicons/core/__init__.py
#! /usr/bin/env python
# -*- coding: utf-8 -*-
"""
A hack to subvert the scons build system for scientific analysis.

Psicons allows scripts and external programs for analysis to be incorporated into
a build procedure.  Every stage of analysis is a command-line call to a script
or executable that takes *inputs* and produces *outputs*. When scons is called, 
dependencies between outputs and inputs are tested and only those stages are run
that are necessary to update outputs. In addition, the exact sequence of 
analyses recorded by the build file.
"""

__docformat__ = 'restructuredtext en'
__version__ = '0.1'
__author__ = "Paul-Michael Agapow (pma@agapow.net)"


### IMPORTS ###

from commands import *
from utils import *


### CONSTANTS & DEFINES ###

### IMPLEMENTATION ###	

## DEBUG & TEST ###

if __name__ == "__main__":
	import doctest
	doctest.testmod()



### END #######################################################################


psicons/core/commands.py
#! /usr/bin/env python
# -*- coding: utf-8 -*-
"""
New scons commands.

The process of installing a new tool or module for use by SCons is fiddly, 
involves copying libraries in a tool directory, registering 
their use in build files, voids the ease of using ``easy_install`` and makes 
development a pain. Thus, these additional "commands" are real scons commands, so
much as functions that generate commands. But they are easy to use.


"""

__docformat__ = 'restructuredtext en'


### IMPORTS ###

import impl


### CONSTANTS & DEFINES ###

### IMPLEMENTATION ###	

def External (env, exe, args=[], infiles=[], output=[], depends=[],
		capture_stdout=None):
	"""
	Create a task for doing analysis with an external (fixed) program.
	
	:Parameters:
		env : Environment
			The environment created for the build
		exe : str
			name of executable to be called
		args : str or list
			either a string that is used to pass arguments to the executable 
			(e.g. ``"--foo 5 -t one.txt"``) or a list of strings that can be joined to
			do the same (e.g. ``["--foo", "5", "-t", "one.txt"]``).
		infiles : str or list
			a string or list of strings giving the paths of the files that this
			task depends on. 
		output : str or list
			a string or list of strings giving the paths of the files that this
			task produces.
		capture_stdout : None or str
			Record the output of the tool. If a string is provided, this will be
			used as the path of the file to save it in.
	
	This is a command that runs an executable, intended for processing input
	datafiles to output data files. It calls the named executable with the
	passed arguments. Infiles and any others listed in `depends` are recorded
	as dependencies. outfiles and the captured output are listed as outputs.
	These will be used in dependency tracking.
	
	"""
	# TODO: should be "outputs"?
	infiles = impl.make_list (infiles or [])
	args = impl.make_list (args or [])
	cmdline = exe % {
		'args': ' '.join (args),
		'infiles': ' '.join (infiles)
	}
	output = impl.make_list (output or [])
	new_command = env.Command (
		output,
		infiles,
		cmdline,
	)
	depends = depends + infiles
	env.Depends (new_command, depends)
	return new_command



def Script (env, path, interpreter=None, args=[], infiles=[], output=[], 
		depends=[], capture_stdout=None):
	"""
	Create a task for running a local script.
	
	:Parameters:
		path : str
			Pathway to the script
		interpreter
			Interpreter to run the script with. If None, assume it is python.
			
		

	This is a command that runs a local script, intended for processing input
	datafiles to output data files. It work like `External`, except that the 
	script itself is made a dependency for this step. Thus ,alterations in the
	script will trigger rerunning. All undescribed parameters are as per `External`.
	
	For example::
	
		>>> env = SconsEnv()
		>>> make_clean_data = Script ('Scripts/clean.py',
			...	args = ['--save-as', 'output.txt'],
			...	infiles = ['indata.txt'],
			...	output = 'output.txt',
		...	)
		>>> type_data = Script ('count_types.pl',
		...		interpreter = 'perl',
		...		args = ['--use-types', 'type_data.csv'],
		...		infiles = ['clean_data.txt'],
		...		depends = 'type_config.txt',
		...		capture_stdout='captured-types.txt',
		...	)
	
	"""
	interpreter = interpreter or '$PY'
	infiles = impl.make_list (infiles or [])
	# TODO: flatten tuples to make arg list like [('-a', 'b'), ('--c', 'd')] to
	# ['-a', 'b', '--c', 'd']?
	# if dict do "--x=foo"
	# TODO: autoescape/quote strings?
	# TODO: have a "SrcFile" class for use in arguments that can be picked up
	# in a scan
	# TOOD: make a "chain" where every script depends on the output of the previous
	args = impl.make_list (args or [])
	cmdline = ' '.join ([interpreter, path] + args + infiles)
	if capture_stdout:
		cmdline = ' '.join ([cmdline, '>', capture_stdout])
		output.append (capture_stdout)
	new_command = env.Command (
		output,
		[path] + infiles,
		cmdline,
	)
	depends = depends + infiles
	env.Depends (new_command, depends)
	return new_command



## DEBUG & TEST ###

if __name__ == "__main__":
	import doctest
	doctest.testmod()



### END #######################################################################



psicons/core/utils.py
"""
Miscellaneous utilities for use in writing and using commands.

Often we will want to name or create a file based on  from pre-existing file
paths:

* create an output file in the same directory as an input file
* clean up some data and save it in a file of the same name with "-clean"
  appended, but before the extension.
* use the same extension for an output file as an input file.
* what is the extension or name of this file anyway?

Thus, here we provide a set of functions for extracting components from a file
path and extrapolating new names from the same. For consistency, here is the
terminology used:

dir
	the path of the directory a file is in
base
	the local name of a file, e.g. "foo.txt", "schedule.doc"
ext
	the extension of a file, e.g. ".txt", ".doc". Note this includes the
	separating period, so that the extensions of files "foo.txt", "foo." and
	"foo" and ".txt", "." and "" respectively.
stem
	the base file name, without the extension, e.g. "foo" and "schedule" for
	"foo.txt", "schedule.doc" respectively.
substem
	the stem of a file name, less the final "word" e.g. "foo" and "schedule_abc"
	for "foo-bar.txt" and "schedule_abc_2011.doc" respectively.

"""

### IMPORTS

import re
from os import path
from datetime import datetime


### CONSTANTS & DEFINES

INTERP_RE = re.compile (r'\{([^}]+)\}')
SUBSTEM_RE = re.compile (r'.([\.\-]\S+)$')


### IMPLEMENTATION

def dir_base (p):
	"""
	Return the directory and file name of a path.
	
	:Parameters:
		p
			a file name or path
			
	:Returns:
		the directory and base name of the file path
	
	For example::
		
		>>> dir_base ('bar.foo')
		('', 'bar.foo')
		>>> dir_base ('/bar.foo')
		('/', 'bar.foo')
		>>> dir_base ('baz/bar.foo')
		('baz/', 'bar.foo')
		>>> dir_base ('/quux/baz/bar.foo')
		('/quux/baz/', 'bar.foo')
		
	We use here the Python convention of referring to the file name (as opposed
	to path) as the "base". Also, there's an apparent oddity in the native
	splitting of paths, to do with the parent directory::
	
		>>> path.split ('bar.foo')
		('', 'bar.foo')
		>>> path.split ('/bar.foo')
		('/', 'bar.foo')
		>>> path.split ('baz/bar.foo')
		('baz', 'bar.foo')
		
	That is, the dividing filesep only appears if there is nothing else in the
	directory component. This makes things difficult if you are trying to pick
	apart a file path and reassemble it - do I need a file separator? was there
	one there before? Fortunately path.join is reasonably clever about fusing
	paths, but this is still an inconsistency and so is fixed here. 
	
	"""
	# TODO: adapt to work with non-Unix file systems (use filesep)
	d, b = path.split (p)
	if (d and (not d.endswith (path.sep))):
		d += path.sep
	return d, b

dirbase_from_path = dir_base

# TODO: rename as split?
def dir_stem_ext (p):
	"""
	Return the directory, stem of the file name and extension from a path.
	
	:Parameters:
		p
			a file name or path
			
	:Returns:
		the directory and base name stem and extension of the file path
	
	For example::
		
		>>> dir_stem_ext ('bar.foo')
		('', 'bar', '.foo')
		>>> dir_stem_ext ('bar.')
		('', 'bar', '.')
		>>> dir_stem_ext ('bar')
		('', 'bar', '')
		>>> dir_stem_ext ('baz/bar.foo')
		('baz/', 'bar', '.foo')
		>>> dir_stem_ext ('/quux/baz/bar.foo')
		('/quux/baz/', 'bar', '.foo')
		
	Note this uses the file separator convention of ``dir_base``.
	
	"""
	dir, base = dir_base (p)
	stem, ext = path.splitext (base)
	return dir, stem, ext

dirstemext_from_path = dir_stem_ext


def substem_mod (s):
	"""
	Split the 'stem' of a file name into a core name and a trailing modifier.
	
	:Parameters:
		s
			a file (base) name
			
	:Returns:
		the substem and modifier (trailing word) of the file name
	
	Many file names are structured as "foo-mod.ext" or "foo.mod.ext" where "mod"
	is some qualifier, e.g. "report-2.doc", "programme.20101201.txt". This
	sniffs out the usual forms of these endings and if found, splits and
	returns the name there.
	
	For example::
	
		>>> substem_mod ('report-2')
		('report', '-2')
		>>> substem_mod ('programme.20101201')
		('programme', '.20101201')
		>>> substem_mod ('report 2')
		('report 2', '')
		>>> substem_mod ('programme_20101201')
		('programme_20101201', '')
		
	"""
	# TODO: what do Macs do with file copies?
	# TODO: what about the windows 'foo (2).txt' pattern? Need multi regexes?
	# XXX: do we need the word separator?
	# XXX: actually would it just be more consistent to split into words and
	#   return [:-1], [-1]
	# TODO: what is there's no end word, only one word?
	# XXX: substem and mod are horrible names. Use words?
	match = SUBSTEM_RE.search (s)
	if (match):
		return s[:match.start(1)], s[match.start(1):]
	else:
		return s, ''


def interpolate (str, sub_hash):
	"""
	Interpolate the bracketed sections of the passed string as keyed substrings.
	
	:Parameters:
		str
			a string containing parens delimited locations for substitution
		sub_hash : dict
			words and substitutions to be used on the string
			
	:Returns:
		the string with substitutions made
		
	The intent for this is as a very simple templating or substitution system to
	be used in configurations and the like. 
	
	For example::
		
		>>> d = {'foo': '123', 'bar': '456'}
		>>> interpolate ('abcdef', d)
		'abcdef'
		>>> interpolate ('abc{foo}def', d)
		'abc123def'
		>>> interpolate ('ab{foo}cd{bar}ef', d)
		'ab123cd456ef'
		
	"""
	# XXX: actually, do we need to do this? are python keywords all that bad?
	#  maybe we should just use a proper templating system?
	# TODO: what does this do with proper python substitutions?
	return INTERP_RE.sub (lambda x: sub_hash[x.group(0)[1:-1].strip()], str)


def interpolate_from_path (p, tmpl, subs={}):
	"""
	Interpolate using qualities of a file path.
	
	:Parameters:
		p : str
			a file name or path
		tmpl : str
			the template for the output, containing parens delimited locations 
			for substitution
		subs : dict
			additional words and substitutions to be used on the string
			
	:Returns:
		the template string with substitutions made
		
	This allows a new file name or path (or actually any string) to be generated
	by interpolation from a file path. This allows construction of paths to files
	in the same directory, files with the same name but different extension, 
	files with the same name except for a suffix, etc.
	
	The substitution keywords are:
	
		ext
			input path file extension, e.g. ".txt"
		base
			input path file base (name), e.g. "foo-bar.txt"
		stem
			input path file name stem, e.g. "foo-bar"
		dir
			input path directory, e.g. "quux/"
		dirstem
			input path directory and stem, e.g. "quux/foo-bar"
		substem
			input path substem, e.g. "foo"
		mod
			input path modifier, e.g. "-bar"
		date
			current date
		time
			current time
			
					
	For example::
		
		>>> pth = '/foo/bar.baz'
		>>> interpolate_from_path (pth, '{stem}.new{ext}')
		'bar.new.baz'
		>>> d = {'prefix': 'PRFX', 'ext': '.txt'}
		>>> interpolate_from_path (pth, '{prefix}{stem}.new{ext}', d)
		'PRFXbar.new.txt'
		
	"""
	# ???: order args as they are used
	# TODO: some sort of number up (tested or untested) or counter
	# TODO: a default output template (so you only need to def subs?)
	_, b = dir_base (p)
	d, s, e = dir_stem_ext (p)
	ss, m = substem_mod (s)
	now = datetime.now()
	default_subs = {
		"ext": e,
		"base": b,
		"stem": s,
		"dir": d,
		"dirstem": d+s,
		"substem": ss,
		"mod": m,
		"date": now.strftime ("%Y%m%d"),
		"time": now.strftime ("%H%M%S"),
		"datetime": now.strftime ("%Y%m%dT%H%M%S"),
	}
	default_subs.update (subs)
	return interpolate (tmpl, default_subs)
	



### MAIN

if __name__ == "__main__":
	import doctest
	doctest.testmod()

### END
